plugins/rest/aws.go

```diff
@@ -318,6 +318,150 @@ func (cs *awsMetadataCredentialService) credentials(ctx context.Context) (aws.Cr
 	return cs.creds, nil
 }
 
+// awsCustomIdentityCredentialService represents an STS credential service that uses active IAM credentials
+// to obtain temporary security credentials generated by AWS STS for allowing existing IAM users access to AWS resources
+// that they don't already have access to.
+type awsCustomIdentityCredentialService struct {
+	RoleArn          string
+	AccessKey        string
+	SecretKey        string
+	SessionToken     string
+	RegionName       string `json:"aws_region"`
+	SessionName      string `json:"session_name"`
+	Domain           string `json:"aws_domain"`
+	stsURL           string
+	creds            aws.Credentials
+	expiration       time.Time
+	logger           logging.Logger
+	signatureVersion string
+}
+
+func (cs *awsCustomIdentityCredentialService) populateFromEnv() error {
+	cs.RoleArn = os.Getenv(awsRoleArnEnvVar)
+	if cs.RoleArn == "" {
+		return errors.New("no " + awsRoleArnEnvVar + " set in environment")
+	}
+
+	cs.AccessKey = os.Getenv(accessKeyEnvVar)
+	if cs.AccessKey == "" {
+		return errors.New("no " + accessKeyEnvVar + " set in environment")
+	}
+	cs.SecretKey = os.Getenv(secretKeyEnvVar)
+	if cs.SecretKey == "" {
+		return errors.New("no " + secretKeyEnvVar + " set in environment")
+	}
+
+	cs.SessionToken = os.Getenv(sessionTokenEnvVar)
+	if cs.SessionToken == "" {
+		// In case of missing SessionToken try to get SecurityToken
+		// AWS switched to use SessionToken, but SecurityToken was left for backward compatibility
+		cs.SessionToken = os.Getenv(securityTokenEnvVar)
+	}
+
+	if cs.Domain == "" {
+		cs.Domain = os.Getenv(awsDomainEnvVar)
+	}
+
+	if cs.RegionName == "" {
+		if cs.RegionName = os.Getenv(awsRegionEnvVar); cs.RegionName == "" {
+			return errors.New("no " + awsRegionEnvVar + " set in environment or configuration")
+		}
+	}
+	return nil
+}
+
+func (cs *awsCustomIdentityCredentialService) stsPath() string {
+	return getSTSPath(cs.Domain, cs.stsURL, cs.RegionName)
+}
+
+func (cs *awsCustomIdentityCredentialService) refreshFromService(ctx context.Context) error {
+	// define the expected JSON payload from the EC2 credential service
+	// ref. https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html
+	type responsePayload struct {
+		Result struct {
+			Credentials struct {
+				SessionToken    string
+				SecretAccessKey string
+				Expiration      time.Time
+				AccessKeyID     string `xml:"AccessKeyId"`
+			}
+		} `xml:"AssumeRoleResult"`
+	}
+
+	// short circuit if a reasonable amount of time until credential expiration remains
+	if time.Now().Add(time.Minute * 5).Before(cs.expiration) {
```

Comment:
Would there be a scenario where a user might want this to be configurable? E.g. if the returned expiration is expected to be close to the 5 min time limit (or maybe these credentials are always expected to live long).

Reply:
We're doing the same for the Web Identity Credentials so I've used the same format. In the future we can surely look into making this configurable if required.